File: scripts/check-nav-v2-read-layer-minimization.mjs
import assert from 'node:assert/strict';
import {
  containsDirectClientIdentifiers,
  maskDealAddress,
  minimizeNavigatorReadPayload,
  neutralDealReference,
  shortDealId
} from '../assets/js/nav-v2/read-layer-minimization-model-v2.js';

const dealId = '12345678-1234-1234-1234-1234567890ab';
const source = {
  profile: {
    id: 'profile-1',
    full_name: 'Рабочий сотрудник',
    phone: '+7 900 000-00-00',
    role: 'spn'
  },
  deal: {
    id: dealId,
    title: 'Квартира — Иванов / Петров',
    display_title: 'Квартира — Иванов / Петров',
    object_type: 'flat_mkd',
    address: 'г. Борисоглебск, ул. Советская, д. 10, кв. 42',
    seller_name: 'Иванов Иван Иванович',
    buyerPhone: '+7 900 111-22-33',
    readiness_deposit: 55,
    wizard_snapshot: {
      sellerPhone: '+7 900 222-33-44',
      buyer_name: 'Петров Пётр Петрович',
      objectFacts: { floor: 3 }
    },
    spn_final: {
      handoff_text: 'Заголовок сделки: Иванов / Петров\nФИО продавца: Иванов И.И.\nСледующий шаг: запросить выписку'
    }
  },
  tasks: [{
    id: 'task-1',
    deal_id: dealId,
    deal_title: 'Иванов / Петров',
    title: 'Запросить выписку ЕГРН',
    description: 'Рабочая задача',
    status: 'open'
  }],
  manager: {
    items: [{
      deal_id: dealId,
      title: 'Иванов / Петров',
      card_url: `./deal-card-v2.html?id=${dealId}`,
      operational_readiness_percent: 40,
      manager_name: 'Алексей Ковтун'
    }]
  },
  mutation: {
    id: 'task-2',
    title: 'Не менять название рабочей задачи',
    status: 'done'
  }
};

const output = minimizeNavigatorReadPayload(source);

assert.notEqual(output, source, 'sanitizer must return a clone');
assert.equal(source.deal.seller_name, 'Иванов Иван Иванович', 'source object must remain unchanged');
assert.equal(containsDirectClientIdentifiers(output), false, 'direct identifiers must be removed recursively');
assert.equal(output.profile.full_name, 'Рабочий сотрудник', 'employee name must remain available');
assert.equal(output.profile.phone, '+7 900 000-00-00', 'employee phone must remain available');
assert.equal(output.deal.address, 'г. Борисоглебск, ул. Советская, д. 10', 'apartment must be removed from address');
assert.equal(output.deal.title, 'Квартира в МКД — г. Борисоглебск, ул. Советская, д. 10 · 12345678');
assert.equal(output.deal.display_title, output.deal.title);
assert.equal(output.deal.wizard_snapshot.objectFacts.floor, 3, 'working facts must survive recursive minimization');
assert.equal(output.deal.spn_final.handoff_text, 'Следующий шаг: запросить выписку', 'legacy identity handoff lines must be removed');
assert.equal(output.tasks[0].title, 'Запросить выписку ЕГРН', 'task title must not be replaced');
assert.equal(output.tasks[0].deal_title, 'Сделка · 12345678', 'deal title embedded in task must be neutral');
assert.equal(output.manager.items[0].title, 'Сделка · 12345678', 'manager queue title must be neutral');
assert.equal(output.manager.items[0].manager_name, 'Алексей Ковтун', 'responsible employee name must remain available');
assert.equal(output.mutation.title, 'Не менять название рабочей задачи', 'non-deal mutation result must not be rewritten');
assert.equal(maskDealAddress('ул. Бланская, 67А, офис 12'), 'ул. Бланская, 67А');
assert.equal(maskDealAddress('ул. Просторная, 4А'), 'ул. Просторная, 4А');
assert.equal(shortDealId('abcdef12-0000'), 'ABCDEF12');
assert.equal(neutralDealReference({ deal_id: dealId }), 'Сделка · 12345678');

console.log('Navigator v2 read-layer minimization semantic regression passed');

File: scripts/check_nav_v2_read_layer_minimization.py
from __future__ import annotations

import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
MODEL = ROOT / "assets/js/nav-v2/read-layer-minimization-model-v2.js"
CLIENT = ROOT / "assets/js/nav-v2/supabase-v2.js"
SEMANTIC = ROOT / "scripts/check-nav-v2-read-layer-minimization.mjs"


def require(text: str, markers: tuple[str, ...], label: str, errors: list[str]) -> None:
    for marker in markers:
        if marker not in text:
            errors.append(f"{label}: missing {marker!r}")


def main() -> int:
    errors: list[str] = []
    for path in (MODEL, CLIENT, SEMANTIC):
        if not path.exists():
            errors.append(f"missing {path.relative_to(ROOT)}")
    if errors:
        print("\n".join(errors))
        return 1

    model = MODEL.read_text(encoding="utf-8")
    require(
        model,
        (
            "import { clientDirectIdentifierKeys }",
            "...clientDirectIdentifierKeys()",
            "export function maskDealAddress",
            "export function neutralDealReference",
            "export function minimizeNavigatorReadPayload",
            "export function containsDirectClientIdentifiers",
            "seller_full_name",
            "buyer_full_name",
            "seller_passport",
            "buyer_passport",
            "seller_snils",
            "buyer_snils",
            "deal_title",
            "dealTitle",
            "handoff_text",
            "Квартира в МКД",
            "Сделка ·",
            "ДЕМО:",
        ),
        MODEL.name,
        errors,
    )

    forbidden = (
        "fetch(",
        "localStorage",
        "sessionStorage",
        "navigator.sendBeacon",
        "XMLHttpRequest",
        "SUPABASE_URL",
    )
    for marker in forbidden:
        if marker in model:
            errors.append(f"{MODEL.name}: pure model contains forbidden side effect {marker!r}")

    client = CLIENT.read_text(encoding="utf-8")
    minimize_call = "let data = minimizeNavigatorReadPayload(await parse(response));"
    recovery_call = "if (name === 'nav_v2_get_deals_list') data = recoverNewDealsOnly(data);"
    require(
        client,
        (
            "import { minimizeNavigatorReadPayload } from './read-layer-minimization-model-v2.js?v=20260716-01';",
            minimize_call,
            "if (name === 'nav_v2_get_my_profile') saveCachedProfile",
            recovery_call,
        ),
        CLIENT.name,
        errors,
    )
    if minimize_call in client and recovery_call in client and client.index(minimize_call) > client.index(recovery_call):
        errors.append("supabase-v2.js: minimization must happen before deal-list recovery and caching")

    semantic = SEMANTIC.read_text(encoding="utf-8")
    require(
        semantic,
        (
            "employee phone must remain available",
            "task title must not be replaced",
            "manager queue title must be neutral",
            "non-deal mutation result must not be rewritten",
            "apartment must be removed from address",
            "demo prefix must survive minimization",
        ),
        SEMANTIC.name,
        errors,
    )

    if errors:
        print("Navigator v2 read-layer minimization contract errors:")
        for error in errors:
            print(f"- {error}")
        return 1

    print("Navigator v2 read-layer minimization contract passed")
    return 0


if __name__ == "__main__":
    sys.exit(main())
